Datum support in Iris

wjbenfold opened the pull request:
## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->

Implements datum support in Iris and hides loading them from NetCDF behind a Future flag.

To do:
- [x] Write a test that checks that we can transform with a datum in the same way as proj does

Reviewer questions:
- I've implemented a small lookup from long datum names to short (proj4 string) ones because it appears impossible to query a proj datum for its short name and a `cartopy.crs.Globe` can't be built from a long one. I would love a better fix for this, any ideas (up to and including changing Cartopy)?
- Are we happy with what's hidden behind the future flag (loading datums from netcdf) and what isn't (everything else)? This seems least likely to break people's code (which would see things move if they've been loading files where we were ignoring the datum before), whilst minimising work required to access datums.
- Should we implement (here or in a separate PR) saving out `crs_wkt` into NetCDF? This can carry datum information, along with other useful information, and is easily pulled from a Cartopy CRS now that they're based on proj. Doing this is currently in the code of this PR but commented out. See also #3796

Reviewer notes:
- I had to change the tests in `lib/iris/tests/unit/fileformats/netcdf/test_Saver.py` because the specified ellipsoids aren't valid and my new implementation had started noticing that.
- Given that every use of `_get_ellipsoid` in `lib/iris/fileformats/_nc_load_rules/helpers.py` is used identically to build the ellipsoid, I just moved the actual building up into `_get_ellipsoid`. This means that currently `unit.fileformats.nc_load_rules.actions.test__grid_mappings.Test__grid_mapping.test_missing_latlon_radius` fails but I don't think that matters so I think that test can be removed.

Closes #4619

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

wjbenfold commented on lib/iris/fileformats/_nc_load_rules/helpers.py:
```diff
@@ -248,21 +252,41 @@ def _get_ellipsoid(cf_grid_var):
         major = getattr(cf_grid_var, CF_ATTR_GRID_EARTH_RADIUS, None)
 
-    return major, minor, inverse_flattening
+    if iris.FUTURE.datum_support:
```
We should warn about future change in behaviour if datum is present but ignored

wjbenfold replied:
done

jamesp commented on lib/iris/fileformats/_nc_load_rules/helpers.py:
```diff
@@ -248,21 +252,41 @@ def _get_ellipsoid(cf_grid_var):
+        datum = None
+
+    if isinstance(datum, str) and (
+        datum.startswith("Unknown") or datum.startswith("unknown")
```
You need a comment here as to why you're catching "unknown"

wjbenfold replied:
[94ffbc5](https://github.com/SciTools/iris/pull/4704/commits/94ffbc5942ebc464539a846047799aae4dca83c8)

wjbenfold commented on lib/iris/fileformats/_nc_load_rules/helpers.py:
```diff
@@ -248,21 +252,41 @@ def _get_ellipsoid(cf_grid_var):
+        datum = None
+
+    if isinstance(datum, str) and (
+        datum.startswith("Unknown") or datum.startswith("unknown")
```
Just handle `=="unknown"` because that's the effective `None` value

wjbenfold replied:
done

jamesp commented on lib/iris/coord_systems.py:
```diff
@@ -303,13 +291,37 @@ def as_cartopy_projection(self):
+            return self._crs.ellipsoid.inverse_flattening
+        if name == "datum":
+            datum = self._crs.datum.name
+            if datum.startswith("unknown") or datum.startswith("Unknown"):
```
see other comment about "unknown" below. maybe this needs to be pulled into a documented subroutine

wjbenfold replied:
[94ffbc5](https://github.com/SciTools/iris/pull/4704/commits/94ffbc5942ebc464539a846047799aae4dca83c8)

jamesp reviewed (changes requested):
@wjbenfold this looks a great improvement.

Before we merge:
- A few small tweaks suggested inline
- As discussed please raise an issue with Cartopy regarding datums and the use of proj4 strings. Let's ascertain whether there is appetite to fix this problem in cartopy (or pyproj) before introducing a datum lookup table here in iris.

wjbenfold commented:
> * Given that every use of `_get_ellipsoid` in `lib/iris/fileformats/_nc_load_rules/helpers.py` is used identically to build the ellipsoid, I just moved the actual building up into `_get_ellipsoid`. This means that currently `unit.fileformats.nc_load_rules.actions.test__grid_mappings.Test__grid_mapping.test_missing_latlon_radius` fails but I don't think that matters so I think that test can be removed.

This was an error - I'd changed the behaviour to fail silently in a case when it shouldn't. Fixed by [89fd422](https://github.com/SciTools/iris/pull/4704/commits/89fd422dae71690c485f8676c96e630fdbc6949b)

wjbenfold commented on lib/iris/fileformats/_nc_load_rules/helpers.py:
```diff
@@ -19,7 +19,9 @@
 import cf_units
 import numpy as np
 import numpy.ma as ma
+import pyproj
```
Done

wjbenfold commented on lib/iris/__init__.py:
```diff
@@ -159,14 +159,14 @@ def __init__(self):
         #
         # self.__dict__['example_future_flag'] = example_future_flag
-        pass
+        self.__dict__["datum_support"] = datum_support
```
[94ffbc5](https://github.com/SciTools/iris/pull/4704/commits/94ffbc5942ebc464539a846047799aae4dca83c8)

wjbenfold commented on lib/iris/__init__.py:
```diff
@@ -159,14 +159,14 @@ def __init__(self):
         # for the structure of this class.
```
Removed the statement that there are none but left the example to reduce future changes

wjbenfold commented on lib/iris/tests/unit/coord_systems/test_LambertConformal.py:
```diff
@@ -74,5 +77,43 @@ def test_optional_args_None(self):
         self._check_crs_defaults(crs)
 
 
+class TestDatumTransformation(tests.IrisTest):
```
done

wjbenfold commented:
Waiting on a check of when we should say the future is coming

wjbenfold commented:
> Waiting on a check of when we should say the future is coming

Conclusion was that we don't tend to say when the future is coming

>Typically we don't speculate when the future will arrive, but if the future is the source of a breaking change, then the future will arrive in the next major release (whenever that is)

So I think we're good to merge @jamesp?

jamesp commented:
Nice work @wjbenfold
